helix-core/src/main/java/org/apache/helix/common/caches/InstanceMessagesCache.java

```diff
@@ -151,6 +155,25 @@ public boolean refresh(HelixDataAccessor accessor, Map<String, LiveInstance> liv
     return true;
   }
 
+  // filter stale message cache by message cache to remove deleted messages
+  public Map<String, Map<String, Message>> getStaleMessageCache() {
+    for (String instanceName : _staleMessageCache.keySet()) {
```

Comment:
I think you can do this when the _messageCache is refreshed in refresh() method.
1. changing values in a get method is not a good style. We want to be able to call get any time, but if it has a side effect then the callers need to review the detail of the method every time if they want to use it.
2. This is for cleaning up and prevent memory leakage only, so clean up when refresh is done at the beginning of the pipeline is enough, I think.